### DISCUSSION TITLE
Make Java a first class citizen for Godot

### DISCUSSION THREAD

**Discussion by eGit** [godotengine/godot-proposals#10772] [General Discussions]
Hello,

there is a huge pool of developers that would prefer using Java to develop with Godot. With the development of Graal and Graal-Native the future of adding Java would be very bright.

Here a short intro and performance test using Java/Kotlin with Godot: https://www.youtube.com/watch?v=oiuECV23lP0

If you watch the video, you can see that Java performance is slightly better than C++.

More performance tests: https://github.com/utopia-rise/godot-kotlin-jvm/tree/master/harness/bunnymark

Here is the project that allows using Java or Kotlin to build games with Godot (currently version 4.2.2): https://godot-kotl.in/en/stable

**Comment by AThousandShips** [comment]
You mention comparisons between java and c++ but the benchmarks in detail don't show anything with c++, would be far more interesting if there was some concrete data and ways to replicate beyond a video

**Reply by eGit** [reply]
The Youtube channel describes the benchmark in an earlier video I think. Maybe there is a link to his Github, but I am sure you can ask for the code in the comments.

Also they converted the "godot-4-3d-third-person-controller" demo to Kotlin: https://github.com/utopia-rise/godot-kotlin-3d-demo

Makes much more sense to just compare the performance using this real world scenarios like this demo.

**Reply by AThousandShips** [reply]
I think it'd be better to remove the claim if it's not easy or possible to test and compare, also it doesn't perform better than c++ according to the video it performs slightly worse, though I couldn't find anything to actually back that up

**Reply by eGit** [reply]
In the video at the end the long running performance-test shows to be better after a while. The JVM is like that. It optimizes more when running longer.

Demo Game in Kotlin/Java/JVM: https://github.com/utopia-rise/godot-kotlin-3d-demo
SAME Demo Game in C++ : https://github.com/gdquest-demos/godot-4-3d-third-person-controller
SAME Demo C++ executable: https://gdquest-demos.itch.io/godot-4-character-3d-demo

Just compare their performance. For me the JVM demo is a smidge faster.

**Reply by AThousandShips** [reply]
What do you mean the same demo in c++ that's in GDScript?

**Comment by AThousandShips** [comment]
I don't think supporting additional languages is something that's going to happen, adding C# was opposed and supported quite strongly as far as I know, and adding even more complexity is a big undertaking and adds a lot of overhead

But regardless this shouldn't be considered until this is production ready, there's plenty of glaring limitations we simply cannot ignore:
* No tool mode (you can set it already in the @RegisterClass annotation but it has no effect yet).
* No plugin support, you cannot use Godot Kotlin/JVM to write plugins and addons yet.
* We support desktop OS (Linux, MacOS, Windows), Android and iOS.

We *already* have the issue of lack of support for C# on web due to upstream limitations, adding another language without support would make Web seem very under prioritized which wouldn't be good

**Comment by dsnopek** [comment]
Having great support for Java/Kotlin in Godot would be awesome!

However, I think it'd be better to focus on making languages added via GDExtensions feel more like first class citizens. There is already some work being done on this as part of the effort to move Godot's C# support to use GDExtension. Then the integration and maintenance of Java/Kotlin support can be done by a sub-community that is interested in that language.

**Comment by eGit** [comment]
youtube.com/watch?v=9AjNzFbTFVo

**Reply by Zireael07** [reply]
1) You posted a top-level reply instead of replying to AThousandShips as you probably intended to

2) It is just a link with no context. For the benefit of others, I can tell you it shows the 3d fps demo game written in Kotlin
